Spigot-Server-Patches/0333-Add-entity-count-cache.patch

```diff
@@ -0,0 +1,172 @@
+index f525fd1b4..494759a1c 100644
+--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
++++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
+@@ -113,7 +113,7 @@ public final class SpawnerCreature {
+                 // CraftBukkit end
+ 
+                 if ((!enumcreaturetype.c() || flag1) && (enumcreaturetype.c() || flag) && (!enumcreaturetype.d() || flag2)) {
+-                    k = worldserver.a(enumcreaturetype.a());
++                    k = worldserver.getCreatureCount(enumcreaturetype); // Paper - entity count cache
+                     int l1 = limit * i / b; // CraftBukkit - use per-world limits
+ 
+                     if (k <= l1) {
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 004c3ec47..dc301c6f4 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -11,6 +11,7 @@ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Random;
+ import java.util.UUID;
++import java.util.EnumMap; // Paper
+ 
+ import java.util.function.Function;
+ import java.util.function.Predicate;
+@@ -50,10 +51,41 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+     // Spigot start - guard entity list from removals
+     public final List<Entity> entityList = new java.util.ArrayList<Entity>()
+     {
++        // Paper start - entity count cache
++        @Override
++        public boolean addAll(Collection<? extends Entity> c) {
++            for (Entity e : c) {
++                updateEntityCount(e, true);
++            }
++
++            return super.addAll(c);
++        }
++
++        @Override
++        public boolean removeAll(Collection<?> c) {
++            for (Object e : c) {
++                if (e instanceof Entity) {
++                    updateEntityCount((Entity)e, false);
++                }
++            }
++
++            return super.removeAll(c);
++        }
++
++        @Override
++        public boolean add(Entity e) {
++            updateEntityCount(e, true);
++
++            return super.add(e);
++        }
++
++        // Paper end
++
+         @Override
+         public Entity remove(int index)
+         {
+             guard();
++            updateEntityCount(get(index), false); // Paper
+             return super.remove( index );
+         }
+ 
+@@ -61,6 +93,7 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         public boolean remove(Object o)
+         {
+             guard();
++            if (o instanceof Entity) updateEntityCount((Entity)o, false); // Paper
+             return super.remove( o );
+         }
+ 
+@@ -2464,6 +2497,53 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
+         return i;
+     }
+ 
++    // Paper start - entity count cache
++    private Map<EnumCreatureType, Integer> countCache = new EnumMap(EnumCreatureType.class);
```

Comment:
Not sure if this is worth it (depends on how often this is updated), but it might be more efficient to use an `int` array here and fill it based on `enum.ordinal()`. That avoids the `Integer` boxing and is pretty much what `EnumMap` does internally anyway.

Reply:
good point

Reply:
hmm so im working on another patch that could *also* use this optimization. was just going to make a static class in EnumCreatureType for this. There any good way to do that in just one patch and use the other as a dependency?

Reply:
Patches are applied sequentially, so if you apply one of the patches first, the others can access stuff the first patch has added. I'm not sure how a static class in `EnumCreatureType` could be of any help here though...

Reply:
wouldnt be huge:

```java
static public class Counter {
        int[] counts = new int[EnumCreatureType.values().length];

        public void incr(EnumCreatureType type) {
            counts[type.ordinal()]++;
        }

        public void decr(EnumCreatureType type) {
            counts[type.ordinal()]--;
        }

        public int get(EnumCreatureType type) {
            return counts[type.ordinal()];
        }
    }
```

can take it or leave it tho 🤷‍♂️

Reply:
That doesn't really look worth it. ;) But in any case, if you need the creature count in another patch as well, why not just reuse the counts you already set up in this patch? :)